File: Codigo Fonte/src/Python/merge_sort.py
def merge(array, indx_esq, meio, indx_dir, mem):
    lado_esq = array[indx_esq : meio + 1]
    lado_dir = array[meio + 1 : indx_dir + 1]

    # Conta os bytes alocados pelos dois slices temporários (8 bytes por ponteiro em CPython 64-bit)
    mem[0] += (indx_dir - indx_esq + 1) * 8

    aux_esq   = 0
    aux_dir   = 0
    aux_array = indx_esq

    while aux_esq < len(lado_esq) and aux_dir < len(lado_dir):
        if lado_esq[aux_esq] <= lado_dir[aux_dir]:
            array[aux_array] = lado_esq[aux_esq]
            aux_esq += 1
        else:
            array[aux_array] = lado_dir[aux_dir]
            aux_dir += 1
        aux_array += 1

    while aux_esq < len(lado_esq):
        array[aux_array] = lado_esq[aux_esq]
        aux_esq += 1
        aux_array += 1

    while aux_dir < len(lado_dir):
        array[aux_array] = lado_dir[aux_dir]
        aux_dir += 1
        aux_array += 1


def merge_sort(array, indx_esq, indx_dir, mem):
    if indx_esq < indx_dir:
        meio = (indx_esq + indx_dir) // 2
        merge_sort(array, indx_esq, meio, mem)
        merge_sort(array, meio + 1, indx_dir, mem)
        merge(array, indx_esq, meio, indx_dir, mem)

File: Codigo Fonte/src/Java/MergeSort.java
import java.util.ArrayList;
import java.util.List;

public class MergeSort {

    void mergeSort(List<Integer> array, int indxEsq, int indxDir, long[] mem) {
        if (indxEsq < indxDir) {
            int meio = (indxEsq + indxDir) / 2;

            mergeSort(array, indxEsq, meio, mem);
            mergeSort(array, meio + 1, indxDir, mem);
            merge(array, indxEsq, meio, indxDir, mem);
        }
    }

    void merge(List<Integer> array, int indxEsq, int meio, int indxDir, long[] mem) {
        int tamanho = indxDir - indxEsq + 1;

        // Conta os bytes alocados pelos dois buffers temporários (4 bytes por int)
        mem[0] += (long) tamanho * 4;

        List<Integer> ladoEsq = new ArrayList<>();
        List<Integer> ladoDir = new ArrayList<>();

        for (int i = indxEsq; i <= meio; i++)
            ladoEsq.add(array.get(i));

        for (int i = meio + 1; i <= indxDir; i++)
            ladoDir.add(array.get(i));

        int auxEsq = 0, auxDir = 0, auxArray = indxEsq;

        while (auxEsq < ladoEsq.size() && auxDir < ladoDir.size()) {
            if (ladoEsq.get(auxEsq) <= ladoDir.get(auxDir)) {
                array.set(auxArray, ladoEsq.get(auxEsq));
                auxEsq++;
            } else {
                array.set(auxArray, ladoDir.get(auxDir));
                auxDir++;
            }
            auxArray++;
        }

        while (auxEsq < ladoEsq.size()) {
            array.set(auxArray, ladoEsq.get(auxEsq));
            auxEsq++;
            auxArray++;
        }

        while (auxDir < ladoDir.size()) {
            array.set(auxArray, ladoDir.get(auxDir));
            auxDir++;
            auxArray++;
        }
    }
}